networkit/cpp/embedding/LearnEmbedding.cpp

```diff
@@ -0,0 +1,229 @@
+        std::generate(emb[i].begin(), emb[i].end(),
+                      [&]() { return (Aux::Random::real() - 0.5) / dimensions; });
+    }
+    return emb;
+}
+
+void trainModel(ModelData &model, count walkNr) {
+    AllWalks &allWalks = model.allWalks;
+    count &wordCntAll = model.wordCntAll;
+    AliasSampler &as = model.as;
+    count allWords = model.allWords;
+    count iterations = model.iterations;
+    count dimensions = model.dimensions;
+    count winSize = model.winSize;
+    double alpha = model.alpha;
+    Embeddings &synPos = model.synPos;
+    Embeddings &synNeg = model.synNeg;
+
+    using Walk = BiasedRandomWalk::Walk;
+
+    Walk &thisWalk = allWalks[walkNr];
+
+    for (index wordI = 0; wordI < thisWalk.size(); ++wordI) {
+        if (wordCntAll % refreshAlphaCount == 0) {
+            alpha = startAlpha * (1 - wordCntAll / ((double)(iterations * allWords) + 1.0));
+            if (alpha < minAlpha) {
+                alpha = minAlpha;
+            }
+        }
+
+        node word = thisWalk[wordI];
+        std::vector<float> eV(dimensions);
+        index offset = Aux::Random::real() * winSize;
+
+        for (index a = offset; a < winSize * 2 + 1 - offset; ++a) {
+            if (a == winSize) {
+                continue;
+            }
+            if (wordI + a < winSize) {
+                continue;
+            }
+            if (wordI + a >= winSize + thisWalk.size()) {
+                continue;
+            }
+            count currWordI = wordI + a - winSize;
+            count currWord = thisWalk[currWordI];
+            // negative sampling
+            for (index j = 0; j < negSamN + 1; ++j) {
+                node target;
+                int label;
+                if (j == 0) {
+                    target = word;
+                    label = 1;
+                } else {
+                    target = as.sample();
+                    if (target == word) {
+                        continue;
+                    }
+                    label = 0;
+                }
+                double product = 0;
+                for (index i = 0; i < dimensions; ++i) {
+                    product += synPos[currWord][i] * synNeg[target][i];
+                }
+                double grad; // Gradient multiplied by learning rate
+                if (product > maxExp) {
+                    grad = (label - 1) * alpha;
+                } else if (product < -maxExp) {
+                    grad = label * alpha;
+                } else {
+                    double exp = std::exp(product);
+                    grad = (label - 1 + 1 / (1 + exp)) * alpha;
+                }
+
+                for (index i = 0; i < dimensions; ++i) {
+                    eV[i] += grad * synNeg[target][i];
+                    synNeg[target][i] += grad * synPos[currWord][i];
+                }
+            }
+
+            for (index i = 0; i < dimensions; ++i) {
+                synPos[currWord][i] += eV[i];
```

Comment:
Are we guaranteed that this walk is the only one writing at this position? If not this must be an atomic increment.